src/passes/DebugLocationPropagation.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// DebugLocationPropagation aim to pass debug location from parents or
+// previous siblings to expression which has no debug location. This is
+// useful for compilers that use Binaryen API to generate WebAssembly modules.
+//
+
+#include "pass.h"
+#include "wasm-traversal.h"
+#include "wasm.h"
+#include <cassert>
+#include <unordered_map>
+
+namespace wasm {
+
+struct DebugLocationPropagation
+  : WalkerPass<PostWalker<DebugLocationPropagation>> {
+  ExpressionStack expressionStack;
```

Comment:
What's the benefit of managing an expression stack yourself instead of using `ExpressionStackWalker` as in the existing code? Also, it looks like this new code does not inherit from previous siblings the way the existing code does. I think it would be good to use the existing code unmodified since we've already verified that it does the right thing.

Reply:
We now using `ExpressionStackWalker` in this pass.

The new code do inherit from previous siblings, the logic is same as existing one. 
And I think we do not need to managing an additional map to store each `expr - debugLocation` pair, because we already store the mapping in Function::debugLocations. So we managed an expression stack to keep track of traversed expressions, the top of the stack is the sibling/parent of current node. The benefit is that we save the memory used by `parentDefaults map` which stored each expression.

Reply:
Ok, I see how this works now. Since you're not actually using the `expressionStack` from `ExpressionStackWalker`, it would actually be better to go back to using `PostWalker`. Can you also add comments to `expressionStack` explaining how it works differently from the `ExpressionStackWalker`'s expression stack so we don't get confused about it in the future?